torch/csrc/autograd/record_function_ops.h

```diff
@@ -0,0 +1,45 @@
+#pragma once
+#include <torch/csrc/autograd/record_function.h>
+#include <torch/csrc/utils/future.h>
+
+namespace torch {
+namespace autograd {
+namespace profiler {
+// Creates a new profiling scope using RecordFunction and invokes its starting
+// callbacks.
+TORCH_API at::Tensor record_function_enter(const std::string& name);
+
+// Cast Tensor that was created with at::cpp_custom_type_hack back to
+// RecordFunction. This is a temporary workaround until RecordFunction is
+// registered as a custom C++ class.
+TORCH_API RecordFunction& getRecordFunctionFromTensor(const at::Tensor& handle);
+
+// Schedules RecordFunction's end callbacks to be run on completion of a future.
+template <typename T>
+TORCH_API void call_end_callbacks_on_fut(
+    const at::Tensor& handle,
+    const std::shared_ptr<torch::utils::Future<T>> fut) {
+  // Add a callback onto the future to mark run RecordFunction's end callbacks
+  // when the future is completed.
+  fut->addCallback(
+      // Copy handle by value to persist after the python context manager is
+      // exited.
+      [handle](
+          const T& /* unused */,
+          const c10::optional<torch::utils::FutureError>& /* unused */) {
+        TORCH_INTERNAL_ASSERT(
+            handle.defined(),
+            "Undefined RecordFunction handle. This can happen if the handle is "
+            "not correctly persisted and is destroyed before the future is "
+            "realized.");
+        auto& rec = getRecordFunctionFromTensor(handle);
+        rec.end();
+      });
+}
+
+// Ends the profiling scope created with record_function_enter.
+TORCH_API void record_function_exit(const at::Tensor& handle);
```

Comment:
I was seeing build errors on OSS and adding these seems to fix it. Although, let me take a look again.

Reply:
I'd really prefer not to expose this as TORCH_API, let's try not to, but at the very least let's mark with _ and add comment that these are internal only and not for the use

Reply:
I was able to remove the TORCH_API on exit and enter, the other 2 functions seem to require them since they are used outside of the cpp.

Reply:
> I was able to remove the TORCH_API on exit and enter, the other 2 functions seem to require them since they are used outside of the cpp.

Which functions are you referring to here? It looks like `record_function_enter` and `record_function_exit` are only used in record_function_ops.cpp, if so we can only keep them in the .cpp file?

Reply:
It looks like `getRecordFunctionFromTensor()` needs to have TORCH_API unless I'm doing something wrong, otherwise I see this error in the OSS build:

```
  File "test/distributed/rpc/test_rpc_spawn.py", line 4, in <module>
    from torch.testing._internal.common_distributed import MultiProcessTestCase
  File "/home/rvarm1/pytorch/torch/__init__.py", line 135, in <module>
    from torch._C import *
ImportError: /home/rvarm1/pytorch/torch/lib/libtorch_python.so: undefined symbol: _ZN5torch8autograd8profiler27getReco
rdFunctionFromTensorERKN2at6TensorE
```

is there a way around this other than adding torch_api?